test/scripts/run_cuda_memcheck.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python
+
+"""This script runs cuda-memcheck on the specified unit test. Each test case
+is run in its isolated process with a timeout so that:
+1) different test cases won't influence each other, and
+2) in case of hang, the script would still finish in a finite amount of time.
+The output will be written to a log file result.log
+
+Example usage:
+    python run_cuda_memcheck.py ../test_torch.py 600
+
+Note that running cuda-memcheck could be very slow.
+"""
+
+import asyncio
+import torch
+import multiprocessing
+import argparse
+import subprocess
+import tqdm
+import re
+
+ALL_TESTS = []
+NUM_PROCESSES = multiprocessing.cpu_count()
+GPUS = torch.cuda.device_count()
+
+# parse arguments
+parser = argparse.ArgumentParser(description="Run isolated cuda-memcheck on unit tests")
+parser.add_argument('filename', help="the python file for a test, such as test_torch.py")
+parser.add_argument('timeout', type=int, help='kill the test if it does not terminate in a certain amount of seconds')
+parser.add_argument('--ignore', nargs='+', default=['cudaErrorInvalidDeviceFunction'],
+                    help='list of regex of the failures not interested, default to ["cudaErrorInvalidDeviceFunction"], '
+                         'because cublas does not run error-free under cuda-memcheck')
+args = parser.parse_args()
+
+# Discover tests:
+# To get a list of tests, run:
+# pytest --setup-only test/test_torch.py
+# and then parse the output
+proc = subprocess.Popen(['pytest', '--setup-only', args.filename], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```

Comment:
If the subprocess segfaults will its output be flushed to the pipe and readable by the other process?

Reply:
(you can trigger a segfault with "import ctypes; ctypes.string_at(0)")

Reply:
You will get something like this in your log file
```
Fail: TestSegFault.testSegFault
========= CUDA-MEMCHECK
========= Error: process didn't terminate successfully
=========        The application may have hit an error when dereferencing Unified Memory from the host. Please rerun the application under cuda-gdb or Nsight Eclipse Edition to catch host side errors.
========= No CUDA-MEMCHECK results found
```

Reply:
Seems reasonable.